### PR TITLE
feat(health): add XState connection health monitor with per-URL state

### LINKED ISSUE #364: Bug: MCP Server startup blocked when GitLab is unreachable
### Description / 问题描述

**English:**
When GitLab is unreachable (network issues, VPN disconnected, GitLab down), MCP Server startup is blocked during `connectionManager.initialize()`. This prevents `tools/list` from returning, causing the MCP client to timeout or fail to load tools.

**中文:**
当 GitLab 不可达（网络问题、VPN 断开、GitLab 宕机）时，MCP Server 在启动过程中被 `connectionManager.initialize()` 阻塞，导致 `tools/list` 无法返回，MCP 客户端超时或工具列表加载失败。

---

### Root Cause / 根因

**English:**
In `handlers.ts`, `connectionManager.initialize()` is called without a timeout during server setup. If GitLab is slow or unreachable, the initialization hangs indefinitely.

**中文:**
在 `handlers.ts` 中，`connectionManager.initialize()` 在服务启动时被调用，但没有设置超时。当 GitLab 响应慢或不可达时，初始化会无限期挂起。

---

### Proposed Fix / 修复方案

**English:**
Add an initialization timeout with `Promise.race`:

```typescript
// config.ts
const parsedInitTimeoutMs = parseInt(process.env.GITLAB_INIT_TIMEOUT_MS ?? "5000", 10);
export const INIT_TIMEOUT_MS =
  Number.isFinite(parsedInitTimeoutMs) && parsedInitTimeoutMs > 0 
    ? parsedInitTimeoutMs : 5000;

// handlers.ts
const initPromise = connectionManager.initialize();
const timeoutPromise = new Promise<never>((_, reject) => {
  setTimeout(
    () => reject(new Error(`Connection initialization timed out after ${INIT_TIMEOUT_MS}ms`)),
    INIT_TIMEOUT_MS
  );
});
await Promise.race([initPromise, timeoutPromise]);
```

**中文:**
添加初始化超时保护：
- 新增 `INIT_TIMEOUT_MS` 配置（默认 5 秒）
- 使用 `Promise.race` 包装初始化操作
- 超时后继续启动，工具列表立即可用，实际调用时再处理连接失败

---

### Benefits / 效果

**English:**
- Server startup completes within 5 seconds regardless of GitLab availability
- Tools list is immediately available
- Actual tool calls handle connection failures gracefully

**中文:**
- 无论 GitLab 是否可用，服务启动最多 5 秒完成
- 工具列表立即可用
- 实际工具调用时优雅处理连接失败

---

### Affected Files / 相关文件

- `src/config.ts` — add `INIT_TIMEOUT_MS` config
- `src/handlers.ts` — wrap `initialize()` with timeout

---

### Environment / 环境

- GitLab MCP version: v6.62.1
- Transport: stdio / HTTP

### PR DESCRIPTION
**Pull request by polaz**
## Summary

- Add per-instance connection lifecycle management using XState v5 state machine (connecting → healthy | degraded | disconnected | failed)
- Auto-reconnect with exponential backoff, bounded startup, disconnected mode with `manage_context`-only tool exposure
- Comprehensive error classification, health probes, URL normalization, and consolidated InstanceRegistry normalization

### Core: `src/services/HealthMonitor.ts` (new)
- XState v5 state machine per instance URL with state callbacks
- `quickHealthCheck()` — unauthenticated HEAD probe with skipAuth + rateLimit bypass, status < 500 = reachable
- `InitializationTimeoutError` class for startup init timeout detection (used in performConnect)
- Handler-level timeouts use plain Error with "timed out" message (distinct from init timeout, classified as transient by classifyError)
- Auto-reconnect: exponential backoff 5s → 60s with ±10% jitter, HEALTH_CHECK_PROBE_MS for fast-path checks

### Connection: `src/services/ConnectionManager.ts`
- Per-URL `InstanceState` via `Map<string, InstanceState>`
- `normalizeInstanceUrl()` applied to all map keys
- Accurate error messages for OAuth-deferred mode (getInstanceInfo/getSchemaInfo)
- resolveState() includes URL in error messages for debugging
- TODO: no-arg callers should pass explicit URL (#379)

### Handlers: `src/handlers.ts`
- Health-aware tool gate: `isInstanceReachable()` for request-scoped URLs
- `bootstrapComplete` set after init + introspection succeed, before cache rebuild
- Late terminal bootstrap errors (auth/permanent) forwarded through `timedOut` guard
- Disconnected `manage_context` fast-path — some actions (e.g. whoami) may attempt API calls that fail gracefully

### Utils
- `normalizeInstanceUrl()` — strip API suffixes, default ports, canonicalize origins via URL constructor
- `parseTimerMs()` — clamp all timer env vars to MAX_SAFE_TIMEOUT_MS
- `skipAuth` option on `enhancedFetch` — suppress auth headers + cookies case-insensitively
- ProxyAgent: `proxyTls.timeout` for per-connection connect budget (not one-shot AbortSignal)
- `classifyError()` — cause.code support, HTTP 408, narrowed timeout heuristic
- `computeDaysUntilExpiry()` — calendar-valid date roundtrip check

### Registry: `src/registry-manager.ts` + `src/services/InstanceRegistry.ts`
- InstanceRegistry.normalizeUrl consolidated with shared `normalizeInstanceUrl` (eliminates port divergence)
- Pool dispatcher lookups use normalized URLs
- Context tools bypass tier/version filtering
- `aggregateFilterCounters()` with reason→counter map
- `postProcessRelatedReferences()` collapsed to single loop
- Per-URL `getFilterStats(instanceUrl?)` — treats connecting as reachable (consistent with isUnreachableMode)

### Config & Docs
- `GITLAB_TOOL_TIMEOUT_MS` (default 120s) documented in README
- Connection resilience section in README.md and README.md.in
- Server keepAliveTimeout clamped to MAX_SAFE_TIMEOUT_MS - 5000 for headersTimeout gap

## Testing

- 4866 tests passing (148 suites)
- HealthMonitor: state transitions, reconnect backoff, health probes, connecting-state isAnyInstanceHealthy
- Handlers: CONNECTION_FAILED states, OAuth URL routing (isConnected/initialize/ensureIntrospected pinned),
  bootstrap timeout with clearInflight, timedOut guard (resolve + reject paths), manage_context health skip,
  connection tracker requestId forwarding
- ConnectionManager: per-URL state, typed helpers, zero `as any`
- URL: default port stripping, API suffix edge cases, malformed URL fallback
- Server: keepAlive clamp boundary test at MAX_SAFE_TIMEOUT_MS
- Build + lint clean

## Related

- #379 — thread instanceUrl through ToolAvailability/WidgetAvailability callers (out of scope)
- #385 — concurrent setupHandlers deduplication test

## Design Notes

**OAuth fallback to GITLAB_BASE_URL**: When OAuth is enabled but no per-request context is available (startup health checks, non-OAuth transports), the handler falls back to `GITLAB_BASE_URL` instead of returning an auth error. This is intentional — blocking health monitoring and tool-list initialization on missing context would degrade the startup experience. The fallback is safe because `GITLAB_BASE_URL` is always configured. Full URL threading is tracked in #379.

## Dependencies

- Yarn 4.12 → 4.13, hono 4.11.7 → 4.12.10 (CVE), all deps bumped

Closes #364

### CODE REVIEW

File: tests/unit/utils/url.test.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    expect(normalizeInstanceUrl('http://gitlab.example.com:80')).toBe('http://gitlab.example.com');
+    // Non-default ports preserved
+    expect(normalizeInstanceUrl('https://gitlab.example.com:8443')).toBe(
+      'https://gitlab.example.com:8443',
+    );
+    // Self-hosted with subpath + default port
+    expect(normalizeInstanceUrl('https://host.com:443/gitlab/api/v4')).toBe(
+      'https://host.com/gitlab',
+    );
+  });
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/unit/utils/url.test.ts]: @coderabbitai Fixed — added test with malformed URLs that exercise the try/catch fallback path.

**Comment by polaz** [comment]
@coderabbitai Re: outside-diff ConnectionManager.test.ts:120-140 — Fixed: mock now forwards args via spread (`...args`) and test asserts `mockDetectTokenScopes` was called with `TEST_URL`.

File: tests/unit/handlers.test.ts [hunk 2]
```diff
@@ -1046,11 +1446,430 @@ describe('handlers', () => {
...
+      // Cover lines 222-223: healthMonitorStartup = null; throw error
+      mockHealthMonitor.initialize.mockRejectedValueOnce(new Error('Health monitor init failed'));
+
+      // First call should throw (propagating the health monitor error)
+      await expect(setupHandlers(mockServer)).rejects.toThrow('Health monitor init failed');
+
+      // After failure, healthMonitorStartup is reset to null so next call retries
+      // A second call should attempt initialization again
+      mockHealthMonitor.initialize.mockResolvedValue(undefined);
+      await expect(setupHandlers(mockServer)).resolves.not.toThrow();
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/unit/handlers.test.ts]: @coderabbitai Fixed — added `expect(mockHealthMonitor.initialize).toHaveBeenCalledTimes(2)` assertion.

**Comment by polaz** [comment]
@coderabbitai Re: outside-diff server.test.ts:95-108 — Added keepAlive clamp boundary test exercising MAX_SAFE_TIMEOUT_MS ceiling. HealthMonitor.shutdown() coverage exists in HealthMonitor.test.ts.

File: tests/unit/handlers.test.ts [hunk 3]
```diff
@@ -1046,11 +1453,431 @@ describe('handlers', () => {
...
+      expect(mockHealthMonitor.initialize).toHaveBeenCalledTimes(2);
+
+      // Restore
+      mockHealthMonitor.initialize.mockResolvedValue(undefined);
+    });
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/unit/handlers.test.ts]: @coderabbitai Fixed — added `expect(mockHealthMonitor.onStateChange).toHaveBeenCalledTimes(1)` after retry.

File: tests/unit/handlers.test.ts [hunk 4]
```diff
@@ -54,27 +59,101 @@ jest.mock('../../src/registry-manager', () => ({
...
 
   beforeEach(() => {
     jest.clearAllMocks();
+    resetHandlersState();
 
     // Create mock server
     mockServer = {
       setRequestHandler: jest.fn(),
-    } as any;
+    } as unknown as jest.Mocked<Server>;
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/unit/handlers.test.ts]: @coderabbitai Won't add tools/list_changed notification assertion — requires SessionManager mock integration which is out of scope for this unit test. The notification path is tested via the broadcastToolsListChangedForStateChange integration in the state-change callback tests.

File: tests/unit/handlers.test.ts [hunk 5]
```diff
@@ -930,10 +1291,51 @@ describe('handlers', () => {
...
+      mockHealthMonitor.reportError.mockClear();
+
+      // Now resolve the delayed tool — this should NOT trigger late health reports
+      resolveDelayed({ result: 'late' });
+      // Give microtasks a chance to settle
+      await new Promise((r) => setTimeout(r, 50));
+
+      expect(mockHealthMonitor.reportSuccess).not.toHaveBeenCalled();
+      expect(mockHealthMonitor.reportError).not.toHaveBeenCalled();
     }, 5000);
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/unit/handlers.test.ts]: @coderabbitai Fixed — replaced `setTimeout(50)` with double `process.nextTick` for deterministic microtask flush.
